src/block/NetherVines.php

```diff
@@ -0,0 +1,183 @@
+		$w->boundedInt(5, 0, self::MAX_AGE, $this->age);
+	}
+
+	public function getAge() : int{
+		return $this->age;
+	}
+
+	/** @return $this */
+	public function setAge(int $age) : self{
+		if($age < 0 || $age > self::MAX_AGE){
+			throw new \InvalidArgumentException("Age must be in range 0-" . self::MAX_AGE);
+		}
+
+		$this->age = $age;
+		return $this;
+	}
+
+	public function isAffectedBySilkTouch() : bool{
+		return true;
+	}
+
+	public function ticksRandomly() : bool{
+		return true;
+	}
+
+	public function canClimb() : bool{
+		return true;
+	}
+
+	abstract protected function getGrowthFace() : int;
+
+	private function getSupportFace() : int{
+		return Facing::opposite($this->getGrowthFace());
+	}
+
+	private function canBeSupportedBy(Block $block) : bool{
+		return $block->getSupportType($this->getSupportFace())->hasCenterSupport() || $block->isSameType($this);
+	}
+
+	public function onNearbyBlockChange() : void{
+		if(!$this->canBeSupportedBy($this->getSide($this->getSupportFace()))){
+			$this->position->getWorld()->useBreakOn($this->position);
+		}
+	}
+
+	private function seekToTop() : NetherVines{
+		$face = $this->getGrowthFace();
+		$top = $this;
+		while(($next = $top->getSide($face)) instanceof NetherVines && $next->isSameType($this)){
+			$top = $next;
+		}
+		return $top;
+	}
+
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		if(!$this->canBeSupportedBy($blockReplace->getSide($this->getSupportFace()))){
+			return false;
+		}
+		$this->age = mt_rand(0, self::MAX_AGE - 1);
+		return parent::place($tx, $item, $blockReplace, $blockClicked, $face, $clickVector, $player);
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($item instanceof Fertilizer){
+			if($this->grow($player, mt_rand(1, 5))){
+				$item->pop();
+			}
+			return true;
+		}
+		return false;
+	}
+
+	public function onRandomTick() : void{
+		if(mt_rand(1, 10) === 1 && $this->age < self::MAX_AGE){
+			if($this->getSide($this->getGrowthFace())->canBeReplaced()){
+				$this->grow(null);
+			}
+		}
+	}
+
+	private function grow(?Player $player, int $growthAmount = 1) : bool{
+		$top = $this->seekToTop();
+		$age = $top->getAge();
+		$pos = $top->getPosition();
+		$world = $pos->getWorld();
+		$growthDirection = $this->getGrowthFace();
+
+		$tx = new BlockTransaction($world);
+
+		for($i = 1; $i <= $growthAmount; $i++){
+			$growthPos = $pos->getSide($growthDirection, $i);
+			if(!$world->isInWorld($growthPos->getFloorX(), $growthPos->getFloorY(), $growthPos->getFloorZ()) || !$world->getBlock($growthPos)->canBeReplaced()){
+				break;
+			}
+			$tx->addBlock($growthPos, (clone $top)->setAge(min(++$age, self::MAX_AGE)));
+		}
+
+		$ev = new StructureGrowEvent($top, $tx, $player);
+		$ev->call();
```

Comment:
The BlockTransaction may be empty. If no blocks are to be generated, the event should not be called.